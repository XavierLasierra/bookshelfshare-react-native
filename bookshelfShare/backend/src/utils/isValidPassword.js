const bcrypt = require("bcrypt");

function isValidPassword(password) {
  return bcrypt.compareSync(password, this.password);
}

module.exports = { isValidPassword };
