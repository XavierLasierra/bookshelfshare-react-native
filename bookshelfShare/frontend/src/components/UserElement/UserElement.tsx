/* eslint-disable no-underscore-dangle */
import React, { useEffect, useState } from 'react';
import {
  Image, Text, View, TouchableOpacity
} from 'react-native';
import { useDispatch } from 'react-redux';

import { addUserFollowing, deleteUserFollowing } from '../../redux/actions/loggedUser.creator';
import { loadCurrentUser } from '../../redux/actions/currentUser.creator';

import RedProfileIcon from '../../assets/redProfileIcon.svg';
import AddUserIcon from '../../assets/addUserIcon.svg';
import styles from './userElement.styles';

export default function UserElement({
  navigation, user, following, loggedUserId, token, refreshToken, logo
}: any) {
  const dispatch = useDispatch();
  const [isDisabled, setIsDisabled] = useState(false);
  const isFollowed = following.some((followedUser: any) => followedUser._id === user._id);

  useEffect(() => {
    setIsDisabled(false);
  }, [following]);

  function handleUserDetail() {
    dispatch(loadCurrentUser(user._id, token, refreshToken));
    navigation.push('OtherUserProfile',
      {
        logo
      });
  }

  function handleAddFollowing(followingUserId: string) {
    dispatch(addUserFollowing(followingUserId, loggedUserId, token, refreshToken));
    setIsDisabled(true);
  }

  function handleDeleteFollowing(followingUserId: string) {
    dispatch(deleteUserFollowing(followingUserId, loggedUserId, token, refreshToken));
    setIsDisabled(true);
  }

  return (
    <>
      {user._id !== loggedUserId && (
      <TouchableOpacity
        style={[styles.userElementContainer, isFollowed && styles.userElementFollowedContainer]}
        onPress={handleUserDetail}
      >
        <Image
          source={{ uri: user.photo }}
          style={styles.userPhoto}
        />
        <View style={styles.nameFollowContainer}>
          <View style={styles.userInformationContainer}>
            <Text style={styles.userName}>{user.username}</Text>
            <Text style={styles.userEmail}>{user.email}</Text>
          </View>
          {isFollowed
            ? (
              <TouchableOpacity
                onPress={() => handleDeleteFollowing(user._id)}
                disabled={isDisabled}
              >
                <RedProfileIcon width={30} height={30} />
              </TouchableOpacity>
            )
            : (
              <TouchableOpacity
                onPress={() => handleAddFollowing(user._id)}
                disabled={isDisabled}
              >
                <AddUserIcon width={30} height={30} />
              </TouchableOpacity>
            )}
        </View>
      </TouchableOpacity>
      )}
    </>
  );
}
