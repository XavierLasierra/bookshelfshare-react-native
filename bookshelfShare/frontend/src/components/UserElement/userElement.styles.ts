import { StyleSheet } from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  userElementContainer: {
    backgroundColor: stylesConstants.colors.white,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: stylesConstants.padding.small,
    borderRadius: 20,
    borderWidth: 1,
    borderColor: stylesConstants.colors.secondary,
    marginVertical: 10
  },
  userElementFollowedContainer: {
    backgroundColor: stylesConstants.colors.follow
  },
  userPhoto: {
    width: 75,
    height: 75,
    borderRadius: 100
  },
  userInformationContainer: {
    width: 150
  },
  userName: {
    fontSize: stylesConstants.fontSize.mid,
    fontWeight: 'bold'
  },
  userEmail: {
    fontSize: stylesConstants.fontSize.mid
  }
});
