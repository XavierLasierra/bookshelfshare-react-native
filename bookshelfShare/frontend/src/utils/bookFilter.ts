export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
