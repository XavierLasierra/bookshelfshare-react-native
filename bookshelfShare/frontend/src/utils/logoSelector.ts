import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
