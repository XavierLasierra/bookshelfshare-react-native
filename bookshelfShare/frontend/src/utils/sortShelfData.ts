export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
