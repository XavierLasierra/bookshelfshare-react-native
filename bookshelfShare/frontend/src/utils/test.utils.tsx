import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
