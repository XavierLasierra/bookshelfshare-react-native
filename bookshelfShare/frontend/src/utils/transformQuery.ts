interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
