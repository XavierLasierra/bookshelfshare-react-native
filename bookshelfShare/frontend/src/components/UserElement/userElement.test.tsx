import React from 'react';
import UserElement from './UserElement';
import {render, fireEvent} from '../../utils/test.utils';
import userSocialsMock from '../../mocks/userSocials.mock';
import {
  addUserFollowing,
  deleteUserFollowing,
} from '../../redux/actions/loggedUser.creator';
import {loadCurrentUser} from '../../redux/actions/currentUser.creator';

jest.mock('../../redux/actions/loggedUser.creator');
jest.mock('../../redux/actions/currentUser.creator');

const navigation = {
  push: jest.fn(),
};

describe('Given a UserElement component', () => {
  describe('When it is rendered', () => {
    describe('And the current user is in following', () => {
      let screen: any;
      beforeEach(() => {
        (deleteUserFollowing as jest.Mock).mockReturnValue({type: ''});
        (loadCurrentUser as jest.Mock).mockReturnValue({type: ''});
        screen = render(
          <UserElement
            navigation={navigation}
            following={userSocialsMock.following}
            user={userSocialsMock.following[0]}
          />,
        );
      });
      test('Then should match the snapshot', () => {
        expect(screen).toMatchSnapshot();
      });

      describe('And deleteFollowingButton is clicked', () => {
        test('Then deleteUserFollowing should be called', () => {
          const deleteFollowingButton = screen.getByTestId(
            'deleteFollowingButton',
          );
          fireEvent.press(deleteFollowingButton);
          expect(deleteUserFollowing).toHaveBeenCalled();
        });
      });

      describe('And userDetailButton is clicked', () => {
        beforeEach(() => {
          const userDetailButton = screen.getByTestId('userDetailButton');
          fireEvent.press(userDetailButton);
        });
        test('Then loadCurrentUser should be called', () => {
          expect(loadCurrentUser).toHaveBeenCalled();
        });
        test('Then navigation.push should be called', () => {
          expect(navigation.push).toHaveBeenCalled();
        });
      });
    });

    describe('And the current user is not in following', () => {
      let screen: any;
      beforeEach(() => {
        (addUserFollowing as jest.Mock).mockReturnValue({type: ''});
        screen = render(
          <UserElement
            navigation={navigation}
            following={[]}
            user={userSocialsMock.following[0]}
          />,
        );
      });
      test('Then should match the snapshot', () => {
        expect(screen).toMatchSnapshot();
      });

      describe('And addFollowingButton is clicked', () => {
        test('Then addUserFollowing should be called', () => {
          const addFollowingButton = screen.getByTestId('addFollowingButton');
          fireEvent.press(addFollowingButton);
          expect(addUserFollowing).toHaveBeenCalled();
        });
      });
    });
  });
});
