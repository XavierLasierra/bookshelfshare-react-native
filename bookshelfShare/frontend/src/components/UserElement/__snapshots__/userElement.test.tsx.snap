// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a UserElement component When it is rendered And the current user is in following Then should match the snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E4FDFF",
      "borderColor": "#ECECEC",
      "borderRadius": 20,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginVertical": 10,
      "opacity": 1,
      "padding": 10,
      "width": 325,
    }
  }
  testID="userDetailButton"
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "borderRadius": 100,
        "height": 75,
        "width": 75,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 200,
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 16,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="deleteFollowingButton"
    >
      <IconMock
        height={30}
        width={30}
      />
    </View>
  </View>
</View>
`;

exports[`Given a UserElement component When it is rendered And the current user is not in following Then should match the snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#ECECEC",
      "borderRadius": 20,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginVertical": 10,
      "opacity": 1,
      "padding": 10,
      "width": 325,
    }
  }
  testID="userDetailButton"
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "borderRadius": 100,
        "height": 75,
        "width": 75,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 200,
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 16,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="addFollowingButton"
    >
      <IconMock
        height={30}
        width={30}
      />
    </View>
  </View>
</View>
`;
